source/utilities.cc

```diff
@@ -0,0 +1,746 @@
+              end_time_dependence (current_file_number, *boundary_id);
+            }
+          else
+            {
+              try
+                {
+                  this->get_pcout() << std::endl << "   Loading Ascii data boundary file "
+                                    << create_filename (next_file_number,*boundary_id) << "." << std::endl << std::endl;
+                  lookups.find(*boundary_id)->second.swap(old_lookups.find(*boundary_id)->second);
+                  lookups.find(*boundary_id)->second->load_file(create_filename (next_file_number,*boundary_id));
+                }
+              catch (...)
+                {
+                  end_time_dependence (current_file_number, *boundary_id);
+                }
+            }
+        }
+    }
+
+
+    template <int dim>
+    std_cxx11::array<unsigned int,dim-1>
+    AsciiDataBoundary<dim>::get_boundary_dimensions (const types::boundary_id boundary_id) const
+    {
+      std_cxx11::array<unsigned int,dim-1> boundary_dimensions;
+
+      switch (dim)
+        {
+          case 2:
+            if ((boundary_id == 2) || (boundary_id == 3))
+              {
+                boundary_dimensions[0] = 0;
+              }
+            else if ((boundary_id == 0) || (boundary_id == 1))
+              {
+                boundary_dimensions[0] = 1;
+              }
+            else
+              AssertThrow(false,ExcNotImplemented());
+
+            break;
+
+          case 3:
+            if ((boundary_id == 4) || (boundary_id == 5))
+              {
+                boundary_dimensions[0] = 0;
+                boundary_dimensions[1] = 1;
+              }
+            else if ((boundary_id == 0) || (boundary_id == 1))
+              {
+                boundary_dimensions[0] = 1;
+                boundary_dimensions[1] = 2;
+              }
+            else if ((boundary_id == 2) || (boundary_id == 3))
+              {
+                boundary_dimensions[0] = 0;
+                boundary_dimensions[1] = 2;
+              }
+            else
+              AssertThrow(false,ExcNotImplemented());
+
+            break;
+
+          default:
+            AssertThrow(false,ExcNotImplemented());
+        }
+      return boundary_dimensions;
+    }
+
+    template <int dim>
+    std::string
+    AsciiDataBoundary<dim>::create_filename (const int filenumber,
+                                             const types::boundary_id boundary_id) const
+    {
+      std::string templ = Utilities::AsciiDataBase<dim>::data_directory + Utilities::AsciiDataBase<dim>::data_file_name;
+      const int size = templ.length();
+      const std::string boundary_name = this->get_geometry_model().translate_id_to_symbol_name(boundary_id);
+      char *filename = (char *) (malloc ((size + 10) * sizeof(char)));
+      snprintf (filename, size + 10, templ.c_str (), boundary_name.c_str(),filenumber);
+      std::string str_filename (filename);
+      free (filename);
```

Comment:
Ugh, this function allocates memory by hand, and it doesn't check the return value of `snprintf`. It's ok for now, but I'm eventually going to rewrite it :-)